flink/v1.12/flink/src/test/java/com/netease/arctic/flink/read/hybrid/enumerator/TestArcticSourceEnumerator.java

```diff
@@ -0,0 +1,253 @@
+          RowKind.INSERT, 1, StringData.fromString("john"), ldt.toEpochSecond(ZoneOffset.UTC), TimestampData.fromLocalDateTime(ldt)));
+        add(GenericRowData.ofKind(
+          RowKind.INSERT, 2, StringData.fromString("lily"), ldt.toEpochSecond(ZoneOffset.UTC), TimestampData.fromLocalDateTime(ldt.plusDays(1))));
+        add(GenericRowData.ofKind(
+          RowKind.INSERT, 3, StringData.fromString("jake"), ldt.toEpochSecond(ZoneOffset.UTC), TimestampData.fromLocalDateTime(ldt.plusDays(2))));
+        add(GenericRowData.ofKind(
+          RowKind.INSERT, 4, StringData.fromString("sam"), ldt.toEpochSecond(ZoneOffset.UTC), TimestampData.fromLocalDateTime(ldt.plusDays(3))));
+      }};
+      for (RowData record : insert) {
+        taskWriter.write(record);
+      }
+      commit(testKeyedTable, taskWriter.complete(), false);
+    }
+  }
+
+  @Test
+  public void testMultipleReaderWhenSplitDiscovered() throws Exception {
+    TestingSplitEnumeratorContext splitEnumeratorContext = instanceSplitEnumeratorContext(5);
+    ShuffleSplitAssigner shuffleSplitAssigner = instanceSplitAssigner(splitEnumeratorContext);
+    ArcticScanContext scanContext =
+      ArcticScanContext.arcticBuilder()
+        .streaming(true)
+        .scanStartupMode(SCAN_STARTUP_MODE_EARLIEST)
+        .build();
+
+    List<ArcticSplit> splitList = FlinkSplitPlanner.planFullTable(testKeyedTable, new AtomicInteger());
+    shuffleSplitAssigner.onDiscoveredSplits(splitList);
+    assertSnapshot(shuffleSplitAssigner, splitCount);
+
+    ArcticSourceEnumerator enumerator = new ArcticSourceEnumerator(
+      splitEnumeratorContext,
+      shuffleSplitAssigner,
+      ArcticTableLoader.of(TableTestHelper.TEST_TABLE_ID, catalogBuilder),
+      scanContext,
+      null,
+      false);
+
+    Collection<ArcticSplitState> pendingSplitsEmpty = enumerator.snapshotState().pendingSplits();
+    Assert.assertEquals(splitCount, pendingSplitsEmpty.size());
+
+    // register readers, and let them request a split
+    // 4 split, 5 subtask, one or more subtask will fetch empty split
+    // subtask 0
+    splitEnumeratorContext.registerReader(0,"host0");
+    enumerator.addReader(0);
+    enumerator.handleSourceEvent(0, new SplitRequestEvent());
+    // subtask 1
+    splitEnumeratorContext.registerReader(1,"host1");
+    enumerator.addReader(1);
+    enumerator.handleSourceEvent(1, new SplitRequestEvent());
+    // subtask 2
+    splitEnumeratorContext.registerReader(2,"host2");
+    enumerator.addReader(2);
+    enumerator.handleSourceEvent(2, new SplitRequestEvent());
+    // subtask 3
+    splitEnumeratorContext.registerReader(3,"host3");
+    enumerator.addReader(3);
+    enumerator.handleSourceEvent(3, new SplitRequestEvent());
+    // subtask 4
+    splitEnumeratorContext.registerReader(4,"host4");
+    enumerator.addReader(4);
+    enumerator.handleSourceEvent(4, new SplitRequestEvent());
+
+    Assert.assertTrue(enumerator.snapshotState().pendingSplits().isEmpty());
```

Comment:
Please also check the `readersAwaitingSplit` size in the enumerator. You can add a getter method with the annotation @VisibleForTesting 
Assert.assertEquals(parallelism - splitCount, enumerator.readersAwaitingSplit.size());